catalog/catalog.go

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package catalog
+
+import (
+	"context"
+	"errors"
+
+	"github.com/apache/iceberg-go/table"
+	"github.com/aws/aws-sdk-go-v2/aws"
+)
+
+type CatalogType string
+
+const (
+	REST     CatalogType = "rest"
+	Hive     CatalogType = "hive"
+	Glue     CatalogType = "glue"
+	DynamoDB CatalogType = "dynamodb"
+	SQL      CatalogType = "sql"
+)
+
+var (
+	// ErrNoSuchTable is returned when a table does not exist in the catalog.
+	ErrNoSuchTable = errors.New("table does not exist")
```

Comment:
I'm not a go expert, but typically we do something like:
```suggestion
	ErrNoSuchTable = errors.New("Table does not exist: {namespace.table_name}")
```

Reply:
@Fokko Yeah I have done this in python, and Java, however in Go it is more common to have these static error types for things like `NotFound` so they can be used to match a specific error. This paired the being forced to handle the error in the same block means you can "wrap" the error at that point and include the table name.

A consumer of the API would do something like this.

```go
table, err := cat.GetTable(identfier)
if err != nil {
    // do something specific on not found
   if errors.Is(err, catalog.ErrNoSuchTable) {
      // something here
   }
   
   return fmt.Errorf("failed to get table (%s): %w", identifier, err)
}
```

Reply:
@wolfeidau Ideally `cat.GetTable` would return `fmt.Errorf("failed to get table (%s): %w", identifier, ErrNoSuchTable)` rather than making the consumer do that wrapping itself.

Reply:
@zeroshade good point, fix added to wrap sentinel not found error.

Reply:
Seems reasonable. Thanks for the explanation 👍